### PR TITLE
AD Powershell Misc Improvements Batch 2

### PR DESCRIPTION
**Pull request by clin-ms**
- Three new checks for the debug cmdlet: KerberosEncryptionType, Channel Encryption, and Domain's Line of Sight. 
- A message at the end of the debug cmdlet output to suggests users to provide more information when reaching out to Azure Files Team
- Better Error message when the user is trying to modify the ad object

### CODE REVIEW

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 1]
```diff
@@ -3148,6 +3148,164 @@ function Debug-AzStorageAccountADObject
...
+        
+        $kerberosTicketEncryptionClient = $adObject.KerberosEncryptionType
+        if(
+            $null -eq $kerberosTicketEncryptionClient -or `
+            0 -eq $kerberosTicketEncryptionClient.Count -or `
+            'None' -eq $kerberosTicketEncryptionClient.Value.ToString()
+            )
+        {
+            Write-Error -Message "No Kerberos Ticket Encryption is supported on the client side" -ErrorAction Stop
+        }
```
Comment (YunZhou-MS) [inline AzFilesHybrid/AzFilesHybrid.psm1:3198]: It's possible that KerberosEncryptionType is not set but a default encryption type is provided by AD. In that case you probably can use klist to get a Keberos ticket and check the Kerberos Encryption Type in the output.

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 2]
```diff
@@ -3148,6 +3148,164 @@ function Debug-AzStorageAccountADObject
...
+
+        if (!$found) 
+        {
+            Write-Error -Message "The server side and the client side do not have a Kerberos Ticket Encryption type in common." -ErrorAction Stop
+        }
+
+    }
+}
+
+function Debug-ChannelEncryption
```
Comment (YunZhou-MS) [inline AzFilesHybrid/AzFilesHybrid.psm1:3232]: This should be AzFileDiagnostics\Windows\AzFileDiagnostics.ps1. Channel encryption is not just for AD scenario. It is a setting for SMB in general.

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 3]
```diff
@@ -3148,6 +3148,164 @@ function Debug-AzStorageAccountADObject
...
+    )
+
+    process
+    {
+
+        $storageAccount = Validate-StorageAccount -ResourceGroupName $ResourceGroupName -StorageAccountName $StorageAccountName -ErrorAction Stop
+
+        $protocolSettings = (Get-AzStorageFileServiceProperty -StorageAccount $storageAccount -ErrorAction Stop).ProtocolSettings.Smb
+
+        $channelEncryptionsClient = (Get-SmbServerConfiguration).EncryptionCiphers.replace("_", "-")
```
Comment (YunZhou-MS) [inline AzFilesHybrid/AzFilesHybrid.psm1:3251]: Need to check if this cmdlet is available for the OS version. If not supported, you probably can use klist to get a Keberos tickert and check the Session Key Type in the output.

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 4]
```diff
@@ -3148,6 +3148,164 @@ function Debug-AzStorageAccountADObject
...
+    )
+
+    process
+    {
+        $storageAccount = Get-AzStorageAccount -ResourceGroupName $ResourceGroupName -Name $StorageAccountName
+        $fullyQualifiedDomainName = $storageAccount.AzureFilesIdentityBasedAuth.ActiveDirectoryProperties.DomainName
+        Write-Host "Fully Qualified Domain Name: $fullyQualifiedDomainName"
+        $checkResult = nltest /dsgetdc:$fullyQualifiedDomainName | Out-String
+
+        if([string]::IsNullOrEmpty($checkResult))
```
Comment (YunZhou-MS) [inline AzFilesHybrid/AzFilesHybrid.psm1:3301]: There will be an error message if nltest fails, so this check is incorrect. You need to parse the output to validate the result.

Comment (clin-ms) [reply AzFilesHybrid/AzFilesHybrid.psm1:3301]: That was my original approach, but I always got NULL as a result when I entered a non-reachable fqdn. I will continue to look into it.

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 5]
```diff
@@ -3148,6 +3148,164 @@ function Debug-AzStorageAccountADObject
...
+        [string]$StorageAccountName,
+
+        [Parameter(Mandatory=$True, Position=1, HelpMessage="Resource group name")]
+        [string]$ResourceGroupName
+    )
+
+    process
+    {
+        $storageAccount = Validate-StorageAccount -ResourceGroupName $ResourceGroupName `
+        -StorageAccountName $StorageAccountName -ErrorAction Stop
```
[the remark this answers is not in the document]
Comment (clin-ms) [reply AzFilesHybrid/AzFilesHybrid.psm1]: fixed

File: AzFilesHybrid/AzFilesHybrid.psm1 (hunk 1 above)
Comment (clin-ms) [reply AzFilesHybrid/AzFilesHybrid.psm1:3219]: Fixed

File: AzFilesHybrid/AzFilesHybrid.psm1 (hunk 3 above)
Comment (clin-ms) [reply AzFilesHybrid/AzFilesHybrid.psm1:3281]: Updated according to the discussion we had: 
- Not able to get ChannelEncryption through Session Key Type, so added extra check to see if the user's system supports  Get-SmbServerConfiguration.

File: AzFilesHybrid/AzFilesHybrid.psm1 (hunk 4 above)
Comment (clin-ms) [reply AzFilesHybrid/AzFilesHybrid.psm1:3331]: $checkResult = nltest /dsgetdc:$fullyQualifiedDomainName | Out-String

After the testing during our discussion, it's proven that nltest's failure message doens't work as the error messages thrown by all other cmdlets. If it fails, we receive null for the result, and not a string with the error message.

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 6]
```diff
@@ -3148,6 +3148,194 @@ function Debug-AzStorageAccountADObject
...
+    {
+
+        $storageAccount = Validate-StorageAccount -ResourceGroupName $ResourceGroupName -StorageAccountName $StorageAccountName -ErrorAction Stop
+
+        $protocolSettings = (Get-AzStorageFileServiceProperty -StorageAccount $storageAccount -ErrorAction Stop).ProtocolSettings.Smb
+
+        $cmdletNeeded = "Get-SmbServerConfiguration"
+        if(!(Get-Command $cmdletNeeded -ErrorAction SilentlyContinue))
+        {
+            Write-Error -Message "Your system does not have or support the command needed for the check '$cmdletNeeded'." -ErrorAction Stop
```
Comment (YunZhou-MS) [inline AzFilesHybrid/AzFilesHybrid.psm1:3278]: Make this a warning and skip CheckChannelEncryption. This is not an error for customer. It is just missing a dependency.

**Review by YunZhou-MS — APPROVED** [review]

**Review by ledavies — APPROVED** [review]

**Pull request merged**